libs/s25main/desktops/dskCampaignSelection.cpp

```diff
@@ -0,0 +1,246 @@
+#include "controls/ctrlText.h"
+#include "controls/ctrlTimer.h"
+#include "dskCampaignMissionSelection.h"
+#include "dskSinglePlayer.h"
+#include "files.h"
+#include "helpers/containerUtils.h"
+#include "helpers/format.hpp"
+#include "ingameWindows/iwMsgbox.h"
+#include "lua/CampaignDataLoader.h"
+#include "ogl/glFont.h"
+#include "gameData/CampaignDescription.h"
+#include "libsiedler2/Archiv.h"
+#include "libsiedler2/ErrorCodes.h"
+#include "libsiedler2/prototypen.h"
+#include "s25util/Log.h"
+#include <boost/filesystem/operations.hpp>
+#include <chrono>
+
+namespace bfs = boost::filesystem;
+using namespace std::chrono_literals;
+
+namespace {
+enum
+{
+    ID_Label,
+    ID_Table,
+    ID_PreviewTitle,
+    ID_PreviewDescription,
+    ID_Back,
+    ID_Next,
+    ID_TimerFillCampaignTable
+};
+
+constexpr int startOffsetY = 20;
+constexpr int padding = 10;
+constexpr int firstColumnExtentX = 480;
+constexpr int secondColumnExtentX = 280;
+constexpr int columnExtentY = 500;
+constexpr int scecondColumnOffsetX = padding * 2 + firstColumnExtentX;
+constexpr int campaignImageExtentY = 220;
+
+int getColumnOffsetY()
+{
+    return startOffsetY + LargeFont->getHeight();
+}
+
+} // namespace
+
+dskCampaignSelection::dskCampaignSelection(CreateServerInfo csi)
+    : Desktop(LOADER.GetImageN("setup015", 0)), showCampaignInfo_(false), csi_(std::move(csi))
+{
+    AddText(ID_Label, DrawPoint(250, startOffsetY), _("Campaigns"), COLOR_YELLOW, FontStyle::CENTER, LargeFont);
+
+    // The table for the campaigns
+    using SRT = ctrlTable::SortType;
+    AddTable(ID_Table, DrawPoint(padding, getColumnOffsetY()), Extent(firstColumnExtentX, columnExtentY),
+             TextureColor::Grey, NormalFont,
+             ctrlTable::Columns{{_("Title"), 100, SRT::String},
+                                {_("Description"), 150, SRT::String},
+                                {_("Author"), 100, SRT::String},
+                                {_("Maps"), 20, SRT::Number},
+                                {_("Difficulty"), 40, SRT::String},
+                                {"", 0, SRT::Default}});
+
+    const int previewTitleOffsetY = getColumnOffsetY() + campaignImageExtentY + padding;
+    AddText(ID_PreviewTitle, DrawPoint(640, previewTitleOffsetY), "", COLOR_YELLOW, FontStyle::CENTER, LargeFont);
+
+    int const multilineOffsetY = previewTitleOffsetY + LargeFont->getHeight();
+    int const mutlilineExtentY = columnExtentY - (multilineOffsetY - getColumnOffsetY());
+    AddMultiline(ID_PreviewDescription, DrawPoint(scecondColumnOffsetX, multilineOffsetY),
+                 Extent(secondColumnExtentX, mutlilineExtentY), TextureColor::Grey, NormalFont);
+
+    AddTextButton(ID_Back, DrawPoint(380, 560), Extent(200, 22), TextureColor::Red1, _("Back"), NormalFont);
+    AddTextButton(ID_Next, DrawPoint(590, 560), Extent(200, 22), TextureColor::Green2, _("Continue"), NormalFont);
+
+    showCampaignInfo(false);
+
+    AddTimer(ID_TimerFillCampaignTable, 1ms);
```

Comment:
Why the timer instead of a direct call?

Reply:
We need the timer here because otherwise the error message showing the campaigns not loaded succesful is not displayed. Because we are in the constructor of the class and the call is not processed.